faiss/gpu/impl/RaftFlatIndex.cu

```diff
@@ -0,0 +1,101 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <faiss/gpu/utils/DeviceUtils.h>
+#include <faiss/gpu/impl/Distance.cuh>
+#include <faiss/gpu/impl/RaftFlatIndex.cuh>
+
+#include <raft/core/device_mdspan.hpp>
+#include <raft/distance/distance_types.hpp>
+//#include <raft/neighbors/brute_force.cuh>
+#include <raft/spatial/knn/detail/fused_l2_knn.cuh>
+
+namespace faiss {
+namespace gpu {
+
+RaftFlatIndex::RaftFlatIndex(
+        GpuResources* res,
+        int dim,
+        bool useFloat16,
+        MemorySpace space)
+        : FlatIndex(res, dim, useFloat16, space) {}
+
+void RaftFlatIndex::query(
+        Tensor<float, 2, true>& input,
+        int k,
+        faiss::MetricType metric,
+        float metricArg,
+        Tensor<float, 2, true>& outDistances,
+        Tensor<idx_t, 2, true>& outIndices,
+        bool exactDistance) {
+    // For now, use RAFT's fused KNN when k <= 64 and L2 metric is used
+    if (k <= 64 && metric == MetricType::METRIC_L2 && vectors_.getSize(0) > 0) {
+        raft::handle_t& raft_handle = resources_->getRaftHandleCurrentDevice();
+
+        auto distance = exactDistance
+                ? raft::distance::DistanceType::L2Unexpanded
+                : raft::distance::DistanceType::L2Expanded;
+
+        auto index = raft::make_device_matrix_view<float>(
+                vectors_.data(), vectors_.getSize(0), vectors_.getSize(1));
```

Comment:
if `FlatIndex::useFloat16_` is true, then the data is stored in `vectorsHalf_` and is of type `half`. Is f32 (queries) x f16 (data) = f32 (restul distance) brute-force kNN supported by RAFT?

Even if f16 is used, we've always been reporting distances as f32 as true nearest neighbors might differ only beyond 2^-10 in the significand fraction.